### PR TITLE
codepush integration

### PR DESCRIPTION
**Pull request by drewvolz**
**CodePush** is a cloud service that enables developers to deploy mobile app updates directly to their users' devices. It works by acting as a central repository that developers can publish updates to (JS, HTML, CSS and images), and that apps can query for updates from.

This allows you to have a more deterministic and direct engagement model with your user base, when addressing bugs and/or adding small features that don't require you to re-build a binary and re-distribute it through the respective app stores.

We will need to redistribute if we add something CodePush is unable to ship out -- something that talks natively to Objective-C/Swift or Java. Something that bridges code. But we are handling new features without bridging new libraries for the most part, so this is a good fit for this project.

It is immaculately documented and provides an outstanding command line tool to help you release an update, rollback to a previous release, or one of the other tasks it allows.

We are going to hit the ground running with 3 lanes/deployments: `nightly`, `staging`, and `release`. We will be able to alpha test, beta test, limited release test, and fully release features with these configurations.

With limited releases, we are able to release features in to the wild on a given percentage of devices. If we are rolling out a new feature but want to be sure we don't witness a lot of new crashes, we can slowly roll something out to 20% of the users, or 50% ... etc. What we roll out in a `nightly` beta could be promoted to the `staging` beta and then introduced into the `release` version.

CodePush also automatically rolls back to a working version of the app once crashes take place, enabling the user to continue using the app without continuously crashing.

A lot can happen with CodePush. For the most part, it enables us to add features and updates _instantly_. No waiting for the app store approval process and no need to wait for users to download a new binary. This is free, this is smart, and this is going to change the way we develop and push features.

<hr>

- [Background information on Microsoft CodePush](https://github.com/Microsoft/react-native-code-push#how-does-it-work)
- [How to use CodePush thru the command line](https://github.com/Microsoft/code-push/blob/master/cli/README.md#codepush-management-cli)

2e7df77c-37bb-11e6-9456-e392af7f7b84

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,6 +16,17 @@
...
+  "codepush": {
+    "ios": {
+      "release": "oSozdf5RtDr2j8T7PFz-oWgNmoBIEJN_nzUOz",
+      "staging": "wU8CTAoyYARqbOBVPXo6kWm7iq8CEJN_nzUOz",
+      "nightly": "lkDky5WyEEILExg-RWX6ymfrJDn6EJN_nzUOz"
+    },
+    "android": {
+      "release": "5jvlQ5-xBItoSRqf-D-3hqdEsPTLEJN_nzUOz",
+      "staging": "GsN8Ii8enIpe8dVijiFQWZ743lV_EJN_nzUOz",
+      "nightly": "cMzHZZ5aUuk0ILftVFFwJvZLxiycEJN_nzUOz"}
```
Comment (hawkrives) [inline package.json]: Please move the closing } to the next line

File: ios/AllAboutOlaf/AppDelegate.m [hunk 2]
```diff
@@ -23,7 +24,12 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
   NSURL *jsCodeLocation;
 
   // see https://github.com/Microsoft/react-native-code-push/issues/393
```
Comment (hawkrives) [inline ios/AllAboutOlaf/AppDelegate.m]: Do we need this comment anymore?

File: android/settings.gradle [hunk 3]
```diff
@@ -1,4 +1,6 @@
 rootProject.name = 'All About Olaf'
+include ':react-native-code-push'
+project(':react-native-code-push').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-code-push/android/app')
```
Comment (hawkrives) [inline android/settings.gradle]: Please alphabetize these below.

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 4]
```diff
@@ -31,6 +38,7 @@ public boolean getUseDeveloperSupport() {
     protected List<ReactPackage> getPackages() {
       return Arrays.<ReactPackage>asList(
         new MainReactPackage(),
+            new CodePush(getResources().getString(R.string.reactNativeCodePush_androidDeploymentKey), getApplicationContext(), BuildConfig.DEBUG),
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java]: Please alphabetize this and move it down and unindent it.

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 5]
```diff
@@ -22,6 +23,12 @@
 public class MainApplication extends Application implements ReactApplication {
 
   private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
+
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java]: Please remove this blank line.

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
 // keep these sorted alphabetically
 import com.facebook.react.ReactApplication;
+import com.microsoft.codepush.react.CodePush;
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java]: Please alphabetize this section. (select the whole chunk; push f5 in sublime)

File: android/app/build.gradle [hunk 7]
```diff
@@ -140,6 +141,7 @@ android {
 }
 
 dependencies {
+    compile project(':react-native-code-push')
```
Comment (hawkrives) [inline android/app/build.gradle]: Please alphabetize this and move it down.

**Review by hawkrives — CHANGES_REQUESTED** [review]
Nitpicks

File: ios/AllAboutOlaf/AppDelegate.m (hunk 2 above)
Comment (drewvolz) [reply ios/AllAboutOlaf/AppDelegate.m]: probably not 👍

**Comment by hawkrives** [comment]
Overall, I'm really excited for CodePush!

**Comment by drewvolz** [comment]
ready to be rebased!

**Review by hawkrives — APPROVED** [review]

**Comment by hawkrives** [comment]
@drewvolz you should probably check to make sure this all works as intended now that I've rebased it.

**Comment by drewvolz** [comment]
@hawkrives confirmed, I've now tested again a deployment to the release versions of iOS and Android.

**Comment by hawkrives** [comment]
@elijahverdoorn There's not a lot to this, but I'd like you to sign off on this before merge.

**Comment by hawkrives** [comment]
Ignore the dependency-ci failure; we disabled it again.

**Comment by drewvolz** [comment]
@elijahverdoorn this is ready for your review.

**Comment by hawkrives** [comment]
For instance, my meals stuff today could have been pushed out through CodePush, or the News thumbnails stuff that's awaiting review.

**Pull request merged**